Fix: SES not emitting SNS messages

simonrw opened the pull request:
Configure SES to emit SNS messages when configured. This is supported in AWS but not currently implemented in Localstack.

This PR builds on #7214.

Note: support for other `Send*Email` methods follows in PRs:

* #7229
* #7228

#7184

bentsku commented on localstack/services/ses/provider.py:
```diff
@@ -196,31 +216,36 @@ def get_identity_verification_attributes(
+                continue
+
+            emitter.emit_send_event(request, sns_destination_arn)
+            emitter.emit_delivery_event(request, sns_destination_arn)
```
Should the events be sent after `call_moto` in case of an exception? I'm not sure, I don't know exactly how SES events work, but maybe they are not sent, or different if something is wrong?

simonrw replied:
@thrau and I were discussing this earlier. I want to spend some time tomorrow trying to see what the error behaviour is for these, since now the services are coupled, errors in either SES or SNS must be handled in the right order. 

In general I think you're right, raising the SES errors first (i.e. sending events after `call_moto`) is the likely right behaviour here anyway :+1:

bentsku commented on localstack/testing/pytest/fixtures.py:
```diff
@@ -1757,6 +1757,68 @@ def _create_rule(**kwargs):
+
+
+@pytest.fixture
+def ses_configuration_set_event_destination(ses_client):
```
There seems to be 4 different destinations possible, maybe this should have `sns` in the name to be aware we would configure only SNS as an event destination?

It also seems SES needs to be allowed to publish to the SNS topic? (not currently enforced in LocalStack, but to be able to AWS-test it)
https://docs.aws.amazon.com/ses/latest/dg/event-publishing-add-event-destination-sns.html
Maybe we could add this policy to the topic while creating the configuration set event destination.

simonrw replied:
That's weird because the test runs (mostly) unmodified on AWS. I only need to change the email addresses and remove the address verification.

simonrw replied:
It seems topics get a default policy that allows the creator wide access:

```json
{
  "Version": "2008-10-17",
  "Id": "__default_policy_ID",
  "Statement": [
    {
      "Sid": "__default_statement_ID",
      "Effect": "Allow",
      "Principal": {
        "AWS": "*"
      },
      "Action": [
        "SNS:GetTopicAttributes",
        "SNS:SetTopicAttributes",
        "SNS:AddPermission",
        "SNS:RemovePermission",
        "SNS:DeleteTopic",
        "SNS:Subscribe",
        "SNS:ListSubscriptionsByTopic",
        "SNS:Publish"
      ],
      "Resource": "arn:aws:sns:eu-central-1:528541226335:test-topic-f5ba40b3",
      "Condition": {
        "StringEquals": {
          "AWS:SourceOwner": "528541226335"
        }
      }
    }
  ]
}
```

Because SES is in the same account, it gets access to `SNS:Publish`

bentsku replied:
Ah, awesome, thanks for checking it!

bentsku reviewed (commented):
Wow, this is awesome! LGTM, I just have some minor nits regarding multi-region and sending the event before calling moto, but otherwise it looks awesome. That's quite the "meatier" PR. 😉 and using snapshots while developing new feature, this is really nice.
Also, by the way, I think this issue would be closed as well by your PR: #5829

simonrw commented on tests/integration/test_ses.py:
```diff
@@ -224,3 +224,114 @@ def test_clone_receipt_rule_set(self, ses_client):
+        queue_arn = sqs_queue_arn(sqs_queue)
+
+        # update queue policy
+        policy = json.dumps(
+            {
+                "Id": f"Policy{short_uid()}",
+                "Version": "2012-10-17",
+                "Statement": [
+                    {
+                        "Sid": f"stmt{short_uid()}",
+                        "Action": "sqs:*",
+                        "Effect": "Allow",
+                        "Resource": queue_arn,
+                        "Condition": {
+                            "ArnEquals": {
+                                "aws:SourceArn": queue_arn,
+                            },
+                        },
+                        "Principal": "*",
+                    },
+                ],
+            }
+        )
+        sqs_client.set_queue_attributes(QueueUrl=sqs_queue, Attributes={"Policy": policy})
+
+        # Subscribe the queue to the topic
+        topic_arn = sns_topic["Attributes"]["TopicArn"]
+        sns_create_sqs_subscription(topic_arn=topic_arn, queue_url=sqs_queue)
```
That's great, thanks!

simonrw commented on localstack/services/ses/provider.py:
```diff
@@ -310,3 +335,81 @@ def clone_receipt_rule_set(
+        client: Optional[SNSClient] = None,
+    ):
+        self.context = context
+        self.client = client if client is not None else aws_stack.connect_to_service("sns")
```
Ok thanks for the pointer, I'll update that :+1:

viren-nadkarni replied:
Correct, also `aws_access_key_id` must be set to the account ID from the ARN. 

There are helpers: `extract_account_id_from_arn` and `extract_region_from_arn`

bentsku replied:
You can also use `parse_arn` in `localstack.utils.aws.aws_stack` and directly extract the `region` and `account` from the `ArnData` typedDict! That's what we used in S3 notifications.

simonrw replied:
Thanks both, I've updated the helper :+1:

bentsku commented on localstack/services/ses/provider.py:
```diff
@@ -196,31 +217,36 @@ def get_identity_verification_attributes(
         )
 
-    @handler("SendEmail")
+    @handler("SendEmail", expand=False)
```
I think the handler is "expanded" in the pro provider, which leads to the test failing with the pro integration because of a `super()` call. Sorry for not picking that up earlier. You might need to use the expanded version.

bentsku reviewed (commented):
Really nice! Thanks for addressing the comments, I just have a last one left because the pro integrations are failing, I've commented the guilty part.

Otherwise, it looks good to me, we might need to emit asynchronously once we add new event destinations, but for now, I'm happy with how it's looking. Thanks a lot!

simonrw commented:
> Really nice! Thanks for addressing the comments, I just have a last one left because the pro integrations are failing, I've commented the guilty part.
> 
> Otherwise, it looks good to me, we might need to emit asynchronously once we add new event destinations, but for now, I'm happy with how it's looking. Thanks a lot!

Thanks - I've made the change to support `-ext` and created two more PRs for extending this to the remaining `Send*Email` methods:

* #7229 implements deletion of configuration sets and configuration set event destinations to properly clean up tests. This is required once running multiple tests that create configuration sets, since they were previously implemented as global state
* #7228 implements the other two methods: `SendTemplatedEmail` and `SendRawEmail`

For the time being they are drafts since they must be merged in order

bentsku commented on localstack/services/ses/provider.py:
```diff
@@ -310,3 +345,96 @@ def clone_receipt_rule_set(
+            Subject="Amazon SES Email Event Notification",
+        )
+
+    @staticmethod
+    def _sender_email(request: SendEmailRequest) -> str:
+        return request["Source"]
+
+    @staticmethod
+    def _destination_addresses(request: SendEmailRequest) -> AddressList:
+        return request["Destination"]["ToAddresses"]
```
nit: are those still in use anywhere since the changes? This is not blocking and can be removed in a follow-up PR!

simonrw replied:
Thanks, yeah I was going to consider that (but TBH I had forgotten! :facepalm: ) I will certainly add this to an upcoming PR since proper error handling is part of the AWS contract :+1:

simonrw replied:
Yeah that's for the future! :tada:

bentsku reviewed (approved):
LGTM! 👍 just a single comment but that can be removed in the follow-up PR you already have open.
Also, I thought about something else as well: what happens if the SNS topic does not exist? What kind of exception does AWS return? We could try several cases, such as:
- When creating the `ConfigurationSetEventDestination` and the topic does not exist, does it validate it? I guess it does by sending the test event. 
- Trying to send an email with an once validated but deleted SNS topic?

But I think this can be part of a subsequent PR, managing exceptions, to not block the following PRs you already have.

All in all, great work! Thanks a lot for adding this functionality and closing 2 SNS related issues! 🚀

simonrw commented:
Thanks, yeah I was going to consider that (but TBH I had forgotten! :facepalm: ) I will certainly add this to an upcoming PR since proper error handling is part of the AWS contract :+1:
